src/DataValueObject.php

```diff
@@ -13,7 +13,20 @@ abstract class DataValueObject implements DataValue {
 	 * @return string
 	 */
 	public function getHash() {
-		return md5( serialize( $this ) );
+		return md5( $this->getLegacySerialization() );
+	}
+
+	/**
+	 * Get legacy PHP serialization (as PHP up to version 7.3 would produce).
+	 * This is used for self::getHash to make sure hashes stay consistent.
+	 * It must not be used to produce serialization meant to be deserialized.
+	 *
+	 * @return string
+	 */
+	public function getLegacySerialization(): string {
```

Comment:
Is this one public just for the test?

If so, using reflection in the test to make it public is probably the lesser evil, as it is local.

Reply:
I need to check Wikibase and see whether we need to have access to this directly (I doubt it though… and if so, we should add this method to the interface, probably).

Reply:
Not all implementations of the interface extend the base class, so moving the new method into the interface would likely not be good.

Example: https://github.com/DataValues/Geo/blob/master/src/Values/GlobeCoordinateValue.php#L116-L123

Which also makes me wonder how many classes do extend `DataValueObject` without overriding `getHash`.

Reply:
Ok, we need direct access to this in `PropertyValueSnak` (as that directly puts the serialization of its DV object into its hash - and not the hash from `DataValueObject::getHash()`).

Reply:
If PHP 7.3- serialization is part of a hash in `PropertyValueSnak`, then the responsibility to generate it seems to be there. Is it possible to create a `Php73SerializationService` thing? That could also be used instead of the new public method, avoiding the need to extend the "interface"/surface area of `DataValueObject`.

Reply:
Well, `EntityIdValue` needs a custom legacy serialization (as it embeds a PHP-serialized entity id… and we want to add `__serialize` to that as well) for `PropertyValueSnak`. As far as I see, we could solve this in the following ways:
1. Add a `getLegacySerialization` method here (and/ or in the interface, which would force everything to implement it), this can then easily be used in `PropertyValueSnak`.
2. Add a `getLegacySerialization` method to `EntityIdValue` (only) and use it in `PropertyValueSnak` (either special cased using `instanceof` or by using `method_exists`).
3. Implement a "full-blown" (to the extend we need it) PHP 7.3-style serialization service

The second option seems like a gigantic code smell and I'm not a fan of the third option, as I don't want to open that can of worms. Therefor I still think we should add this here (and in the interface?).

Reply:
Note: I am genuinely curious here and really do not wish to be drawn in a Wikimedia style discussion filled with rhetoric.

> Implement a "full-blown" (to the extend we need it) PHP 7.3-style serialization service

You make it seem like this would be a big deal. Would such a class not be simple and easy to create? Not much more than the lines of code you already added here?

> The second option seems like a gigantic code smell

Adding a new public method in this library because of something in EntityIdValue and PropertyValueSnak is also suspect. That is why I hope there is a better solution than the current PR. Is there no way to solve the issue in these classes without resorting to needing to check for the instance type?

Reply:
As long as `getLegacySerialization` is not added to the interface, to me option 1 seems fairly similar to option 2: `PropertyValueSnak` should check if the method is available either way, because it’s supposed to work with arbitrary implementations of the `DataValue` interface, not just with `DataValueObject` subclasses.

And as long as we use `method_exists()` rather than `instanceof` in option 2, and maybe choose a different method name (`getSerializationForHash` sounds nicer than “legacy” to me), then I think this is an acceptable option: more dynamic than the PHP code we usually write, sure, but not terrible, considering we’re still working in a dynamic language. So I  would advocate for that option.

Also, just to make sure I understand correctly: strictly speaking, we only need `getLegacySerialization()` here, but the `hash()` shouldn’t actually matter, right? Since what we really care about is the //snak// hash, which is based on the data value’s serialization, not its hash. (Though I feel like, especially if we change the method name to something like `getSerializationForHash()`, we should still override `hash()` here. It would be weird to have a `getSerializationForHash()` that’s then not used by the class’ own `hash()`.)

Reply:
Is the a concrete problem in Wikibase if the hashes of the data values change?

If not, then we could also use this as an opportunity to have getHash not use serialize (like done in the globe values https://github.com/DataValues/Geo/blob/master/src/Values/GlobeCoordinateValue.php#L116-L123).